### PR TITLE
Octane 5.7.1 fixes

### PR DESCRIPTION
**Pull request by radislavB**
[JENKINS-56360 : Sonar-octane Integration - add support for pipeline as a code](https://issues.jenkins-ci.org/browse/JENKINS-56360)
[JENKINS-56361 : JGIT-octane integration - enrich line changes on SCM commits](https://issues.jenkins-ci.org/browse/JENKINS-56361)
[JENKINS-56433 : Removing temp file GetJUnitTestResultsXXXXXXXXXXXXXXX.tmp from job build](https://issues.jenkins-ci.org/browse/JENKINS-56433)

### CODE REVIEW

File: src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.microfocus.application.automation.tools.octane;
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java]: Missing license

File: src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import com.microfocus.application.automation.tools.octane.configuration.ConfigurationService;
+import hudson.model.User;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Collections;
+
+public class ImpersonationUtil {
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java]: Please add a description, on the purpose of this new class.

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java [hunk 3]
```diff
@@ -467,27 +465,11 @@ public OctaneResponse upsertCredentials(CredentialsInfo credentialsInfo) {
...
-			if (jenkinsUser == null) {
-				throw new PermissionException(401);
-			}
-		} else {
-			logger.info("No user set to impersonating to. Operations will be done using Anonymous user");
-		}
-
-		ACLContext impersonatedContext = ACL.as(jenkinsUser);
-		return impersonatedContext;
+		return ImpersonationUtil.startImpersonation(getInstanceId());
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: If an exception occurs, it won't be caught?

File: src/main/java/com/microfocus/application/automation/tools/octane/actions/Webhooks.java [hunk 4]
```diff
@@ -105,68 +107,72 @@ public void doNotify(StaplerRequest req, StaplerResponse res) throws IOException
...
-                                String sonarToken = SonarHelper.getSonarInstallationTokenByUrl(sonarConfiguration, action.getServerUrl());
-                                HashMap project = (HashMap) inputNotification.get(PROJECT);
-                                String sonarProjectKey = (String) project.get(SONAR_PROJECT_KEY_NAME);
-
-                                if (action.getDataTypeList().contains(SonarHelper.DataType.COVERAGE)) {
-                                    // use SDK to fetch and push data
-                                    OctaneSDK.getClients().forEach(octaneClient -> octaneClient.getSonarService().enqueueFetchAndPushSonarCoverage(jobName, buildId, sonarProjectKey, action.getServerUrl(), sonarToken));
+                for (OctaneClient octaneClient : OctaneSDK.getClients()) {
+                    String octaneInstanceId = octaneClient.getInstanceId();
+                    ACLContext aclContext = ImpersonationUtil.startImpersonation(octaneInstanceId);
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/octane/actions/Webhooks.java]: Same

File: src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java [hunk 5]
```diff
@@ -119,25 +134,23 @@ public boolean isPushCoverage() {
...
 				if (allConfigurations != null) {
 					SonarHelper adapter = new SonarHelper(run, listener);
-					serverDetails[0] = sonarServerUrl.isEmpty() ? adapter.getServerUrl() : sonarServerUrl;
-					serverDetails[1] = sonarToken.isEmpty() ? adapter.getServerToken() : sonarToken;
+					setSonarServerUrl(sonarServerUrl.isEmpty() ? adapter.getServerUrl() : sonarServerUrl);
+					setSonarToken(sonarToken.isEmpty() ? adapter.getServerToken() : sonarToken);
 
 				}
 			} catch (Exception e) {
 				throw new InterruptedException("exception occurred while init sonar tracker for job " + run.getDisplayName() + " error message: " + e.getMessage());
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java]: If the intention of this is failing the build, I am not sure if it's the right approach.

File: src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java [hunk 6]
```diff
@@ -161,31 +174,27 @@ private String getBuildNumber(Run<?, ?> run) {
...
+		run.addAction(new WebhookAction(true, getSonarServerUrl(), dataTypeSet));
 	}
 
 	@Override
 	public SonarDescriptor getDescriptor() {
 		return (SonarDescriptor) super.getDescriptor();
 	}
 
-
+	@Symbol("addSonarQubeListener")
```
Comment (grount) [inline src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java]: The name of the pipeline will be "addSonarQubeListener" with no reference to Octane, wouldn't it be misleading to SonarQube Jenkins plugin?

File: src/main/java/com/microfocus/application/automation/tools/octane/model/processors/scm/GitSCMProcessor.java [hunk 7]
```diff
@@ -71,7 +78,166 @@
...
+                    }
+                }
+                fileBlameList.add(new SCMFileBlameImpl(filePath, revisionsMap));
+
+            }
+
+            return fileBlameList;
+
+        } catch (IOException | GitAPIException e) {
+            throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (grount) [reply src/main/java/com/microfocus/application/automation/tools/octane/model/processors/scm/GitSCMProcessor.java]: Why is RuntimeException occurs?

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java (hunk 3 above)
Comment (eemanuelhpe) [reply src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: the only change that I done here is to move "startImpersonation" method to different class. but basically, this is very old code, which been used in a lot of different contexts. so I don't want to change it now.
so for now, I caught the exception (in the fix to PR) only when I used this method, in Webhooks class.

File: src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java (hunk 1 above)
Comment (radislavB) [reply src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java]: Added

File: src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java (hunk 5 above)
Comment (radislavB) [reply src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java]: This code is not running as part of build but as thread after job is finished. Its not intended to fail the build.

File: src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java (hunk 2 above)
Comment (radislavB) [reply src/main/java/com/microfocus/application/automation/tools/octane/ImpersonationUtil.java]: Added

File: src/main/java/com/microfocus/application/automation/tools/octane/actions/Webhooks.java (hunk 4 above)
Comment (radislavB) [reply src/main/java/com/microfocus/application/automation/tools/octane/actions/Webhooks.java]: see comment of Eli

File: src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java (hunk 6 above)
Comment (radislavB) [reply src/main/java/com/microfocus/application/automation/tools/run/SonarOctaneListener.java]: Changed to addALMOctaneSonarQubeListener

**Pull request merged**